lib/src/clojure_lsp/parser.clj

```diff
@@ -27,54 +27,65 @@
        (mapv (fn [loc] (z/find loc z/up edit/top?)))
        (distinct)))
 
+(def ^:private zero-width-space
+  "A unicode character that is incredibly unlikely to be used in regular code.
+  During parsing, used as a valid and easily identified subsitute for what would
+  otherwise be an invalid character."
+  "\u200b")
+
+(defn ^:private replace-incomplete-token [s invalid-str valid-str]
+  (let [token-pattern (re-pattern (str invalid-str "(\\s|\\n|\\))"))]
+    (when-let [[_ divider] (re-find token-pattern s)]
+      (string/replace-first s token-pattern (str valid-str divider)))))
+
+(defn ^:private z-replace-preserving-meta [zloc replacement]
+  (z/replace zloc (with-meta replacement (meta (z/node zloc)))))
+
 (defn ^:private handle-end-slash-code [text exception]
   (when-let [[_ token] (->> exception
                             Throwable->map
                             :cause
-                            (re-matches #"Invalid symbol: (.*\/)."))]
-    (let [token-pattern (re-pattern (str token "(\\s|\\n|\\))"))]
-      (when-let [[_ extra-token] (re-find token-pattern text)]
-        (-> text
-            (string/replace-first token-pattern (str token "_" extra-token))
-            z/of-string
-            (z/edit->
-              (z/find-next-value z/next (symbol (str token "_")))
-              (z/replace (n/token-node (symbol token))))
-            z/up)))))
+                            (re-matches #"Invalid symbol: (.*)\/."))]
+    (let [real-value      (str token "/")
+          temporary-value (str token zero-width-space)]
+      (some-> text
+              (replace-incomplete-token real-value temporary-value)
+              z/of-string
+              (z/edit->
+                (z/find-value z/next (symbol temporary-value))
+                (z-replace-preserving-meta (n/token-node (symbol real-value))))))))
 
 (defn ^:private handle-single-colon-code [text exception]
   (let [cause (->> exception Throwable->map :cause)]
     (when (or (re-matches #"\[line (\d+), col (\d+)\] A single colon is not a valid keyword." cause)
               (re-matches #"\[line (\d+), col (\d+)\] Invalid keyword: ." cause))
-      (let [colon-pattern (re-pattern ":(\\s|\\n|\\))")]
-        (when-let [[_ extra-token] (re-find colon-pattern text)]
-          (-> text
-              (string/replace-first colon-pattern (str ":___" extra-token))
-              z/of-string
-              (z/edit->
-                (z/find-next-value z/next :___)
-                (z/replace (n/token-node (symbol ":"))))
```

Comment:
One of the biggest workarounds I did on clojure-lsp I think, glad it's going away 🎉

Reply:
Oh, it's not going away, but you improved the code, nice

Reply:
Yeah, not going away, just a refactoring to make sure that incomplete nodes and their neighbors get position metadata. Actually, it's a pretty elegant hack. The [tree-sitter docs](https://tree-sitter.github.io/tree-sitter/#underlying-research) talk about how to make parsers that can recover from errors, which use a similar technique.